### PR TITLE
Wasm web support

### PR DESCRIPTION
**Pull request by Anderssorby**
This adds:

- A wasm feature
- The ability to compile without fd-lock which makes it possible to compile to wasm

### CODE REVIEW

**Comment by gwenn** [comment]
https://doc.rust-lang.org/cargo/guide/cargo-toml-vs-cargo-lock.html
> If you’re building a non-end product, such as a rust library that other rust [packages](https://doc.rust-lang.org/cargo/appendix/glossary.html#package) will depend on, put Cargo.lock in your .gitignore.

**Comment by johnchandlerburnham** [comment]
@gwenn Fixed. Tests pass locally for me, but if you can approve us to run workflows we can go through CI.

Also happy to keep/remove the Nix build from this PR (can move it to a separate PR if its of interest)

**Comment by gwenn** [comment]
https://github.com/nushell/nushell/pull/830/commits/65008bb912076f288d0c0dbf8ba0ae9f3641d374 ?
And are you sure that `nix` files are useful for a rust library ?

**Comment by johnchandlerburnham** [comment]
Not sure what `nushell` has to do with it, but there are advantages to having a nix build, such as portability for contributors. However, it also adds some complexity which you might not want.

Happy to keep or remove the nix build, up to you.

**Comment by gwenn** [comment]
And there is a related pending PR:
https://github.com/kkawakam/rustyline/pull/604/files#diff-2e9d962a08321605940b5a657135052fbcef87b5e360662bb527c96d9a615542
where `FileHistory` can be made optional.

**Comment by Anderssorby** [comment]
@gwenn I guess it should be easy to integrate this into your PR?

**Comment by gwenn** [comment]
Except it is redundant with PR #604

**Comment by johnchandlerburnham** [comment]
It's not redundant. This PR is a minimal change of 25 lines which gets `rustyline` building on WASM. #604 is a >700 line change feature-level change which has been open for 2 months and has 2 of 5 tasks completed. It seems like merging this PR should be much easier, which will then let us use rustyline for our web repl: https://github.com/lurk-lang/lurk-rs/pull/57

That said, if you guys prefer to wait for #604 to have wasm support we can just fork off a separate crate for our purposes in the meantime.

File: Cargo.toml [hunk 1]
```diff
@@ -58,11 +58,12 @@ assert_matches = "1.2"
 rustyline-derive = { version = "0.6.0", path = "rustyline-derive" }
 
 [features]
-default = ["custom-bindings", "with-dirs"]
+default = ["custom-bindings", "with-dirs", "fd-lock"]
 custom-bindings = ["radix_trie"]
 with-dirs = ["dirs-next"]
 with-fuzzy = ["skim"]
 case_insensitive_history_search = ["regex"]
+wasm = [ ]
```
Comment (gwenn) [inline Cargo.toml:66]: Useless ?

File: src/history.rs [hunk 2]
```diff
@@ -170,6 +172,12 @@ impl History {
         self.update_path(path, &lock_guard, self.len())
     }
 
+    #[cfg(not(feature = "fd-lock"))]
+    pub fn save<P: AsRef<Path> + ?Sized>(&mut self, path: &P) -> Result<()> {
+        // TODO
+        Ok(())
```
Comment (gwenn) [inline src/history.rs:178]: `unimplemented!()` / `todo!()`

File: src/history.rs [hunk 3]
```diff
@@ -251,10 +260,17 @@ impl History {
         Ok(())
     }
 
+    #[cfg(not(feature = "fd-lock"))]
+    pub fn append<P: AsRef<Path> + ?Sized>(&mut self, path: &P) -> Result<()> {
+        // TODO
```
Comment (gwenn) [inline src/history.rs:265]: `unimplemented!()` / `todo!()`

**Comment by gwenn** [comment]
- [ ] fix docs.rs [features](https://github.com/kkawakam/rustyline/blob/master/Cargo.toml#L68)

**Comment by Anderssorby** [comment]
@gwenn I've fixed the comments and warnings. Let me know if there is anything else you want to change.

**Comment by Anderssorby** [comment]
@gwenn Any updates on this? It would be awkward to have to crate a new crate just to fix wasm compilation...

**Comment by Anderssorby** [comment]
I feel this is more explicit and meaningful. Maybe someone adds wasm (with wasi) support for this? Also one can now explicitly disable the fd-lock dependency for other cases too.

**Comment by gwenn** [comment]
> Also one can now explicitly disable the fd-lock dependency for other cases too.
You mean like in PR #604 !

**Comment by samuelburnham** [comment]
Now that #604 is merged, rustyline compiles to `wasm32-unknown-unknown` on the master branch when the `with-file-history` feature is disabled. Thanks @gwenn! There are still a few warnings about unused variables/imports, but I think it's safe to close the PR.

**Pull request closed without merging**